File: 혜원/W6_Sort/파일명_정렬.py
# 적용 알고리즘의 개념
    # 정규표현식 + 정렬
    # 파일명에서 NUMBER를 먼저 뽑아 정렬하고,
    # 그 다음 HEAD를 기준으로 다시 정렬하는 방식

# 문제 풀이를 위한 접근 방식(개념) 설명
    # 정규표현식을 이용해서
    # 숫자 부분(NUMBER)과 문자 부분(HEAD)을 쉽게 구할 수 있다.

    # 1. 먼저 NUMBER 기준으로 정렬하고,
    # 2. 그 다음 HEAD 기준으로 다시 정렬
    # 3. 파이썬의 sorted는 안정 정렬이기 때문에, HEAD가 같은 경우 앞에서 정렬한 NUMBER 순서가 그대로 유지된다.

# 기본 코드와 개선한 코드 비교
    # 직접 HEAD와 NUMBER를 나누는 방법도 가능하지만,
    # 이 코드는 re를 사용해서 더 짧고 간단하게 구현한 풀이이다. 
    # O(n log n)

# 사용 라이브러리 정리
    # re.findall(r'\d+', file)[0]
        # 파일명에서 처음 나오는 숫자 부분 추출
    # re.split(r'\d+', file.lower())[0]
        # 숫자를 기준으로 나눠서 HEAD 부분 추출
        # lower()를 사용해서 대소문자 구분 없이 비교

import re

def solution(files):
    # 먼저 NUMBER 기준으로 정렬
    arr = sorted(files, key=lambda file: int(re.findall(r'\d+', file)[0]))

    # 그 다음 HEAD 기준으로 정렬
    arr = sorted(arr, key=lambda file: re.split(r'\d+', file.lower())[0])

    return arr

File: 명찬/W6/파일명 정렬.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <tuple>
using namespace std;

int type_check(char a) // 숫자 0, 소문자 1, 대문자 2, 기호 3
{
    if ('0' <= a && a <= '9')
    {
        return 0;
    }
    else if ('a' <= a && a <= 'z')
    {
        return 1;
    }
    else if ('A' <= a && a <= 'Z')
    {
        return 2;
    }
    else
    {
        return 3;
    }
}

vector<string> solution(vector<string> files) {
    vector<string> answer;
    vector<tuple<string, int, int>> v; //헤드와 숫자를 분리하여 저장해놓을 배열
    for (int i = 0; i < files.size(); i++) //파일 순회
    {
        string name = files[i];
        string head = ""; //헤드쪽 저장
        string number = ""; //숫자 저장
        int state = 0; //현재 상태. 헤드 수집중이면 0, 숫자 수집중이면 1
        for (int j = 0; j < name.size(); j++)
        {
            char t = name[j]; //파일명의 한 글자
            int type_num = type_check(t); //타입체크. 숫자 0, 소문자 1, 대문자 2, 기호 3
            if (state == 0) //head 수집단계
            {
                if (type_num == 1 || type_num == 3) //소문자나 기호라면
                    head.push_back(t);
                else if (type_num == 2) //대문자라면
                    head.push_back(t - 'A' + 'a'); //대소문자 통합
                else if (type_num == 0) //숫자라면
                {
                    state = 1; //number 수집 단계로 이동
                    number.push_back(t);
                }
            }
            else if (state == 1) //number 수집 단계
                if (type_num == 1 || type_num == 2 || type_num == 3) //숫자 이외의 것이라면
                    break; //탈출
                else if (type_num == 0)
                    number.push_back(t);
        }
        v.push_back({ head,stoi(number),i }); //헤드, 숫자, 순서
    }
    sort(v.begin(), v.end());
    for (int i = 0; i < v.size(); i++)
        answer.push_back(files[get<2>(v[i])]); //순서를 읽어서 파일 인덱스로 텍스트 가져와 push_back
    return answer;
}